### PR TITLE
Fix git ref to allow branches and tags

### LINKED ISSUE #2932: Git inputs cannot reference tags with the ref option
As far as I understand, git branches and tags are types of git refs (they are stored under `.git/refs`).

IMO it should therefore be possible to use the `ref` option as a generic way for referencing commits, branches, and tags.

This is actually stated in the documentation:

>Any ref that is a valid input to git checkout is accepted.

https://buf.build/docs/reference/inputs#other-options

However, while `git checkout remote-tag` is working, `buf generate "ssh://git@github.com/path-to-repo.git#ref=remote-tag"` is not.

IMO this feature would make parameterizing CI pipelines easier, e.g.: 

- `ssh://git@github.com/path-to-repo.git#ref=develop` - build against latest changes on the develop branch
- `ssh://git@github.com/path-to-repo.git#ref=v1.0.0` - build against a stable release tag

Note: this is also how kubectl kustomize is implementing the `ref`option:

> ref - a [git fetch-able ref](https://git-scm.com/docs/git-fetch), typically a branch, tag, or full commit hash (short hashes are not supported)

https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md#remote-directories

### PR DESCRIPTION
**Pull request by emcfarlane**
This changes the git ref input to align with the git notion of a ref as a tag or branch. We now always try to fetch a given ref before falling back to fetching the HEAD to resolve partial refs. If both a ref and branch are provided, the branch is fetched directly before resolving the ref. The inputs `tag` and `commit` can still be used but can now be replaced by setting the `ref`.

For example:
 - Branch `ref=main` will fetch and checkout the `main` branch
 - Tag `ref=v1.0.0` will fetch and checkout the `v1.0.0` tag
 - Partial `ref=123456` will fetch `HEAD` and checkout `123456`
 - Branch with partial `ref=123456,branch=feature` will fetch `feature` and checkout `123456`
 - Commit `ref=<sha>` will fetch `<sha>` and checkout `<sha>`
 - Branch with commit `ref=<sha>,branch=feature` with fetch `feature` and checkout `<sha>`

Fixes #2932

### CODE REVIEW

File: private/pkg/git/cloner.go [hunk 1]
```diff
@@ -97,51 +88,38 @@ func (c *cloner) CloneToBucket(
...
-	defer func() {
-		retErr = multierr.Append(retErr, worktreeDir.Close())
+		retErr = multierr.Append(retErr, baseDir.Close())
 	}()
 
 	buffer := bytes.NewBuffer(nil)
 	if err := c.runner.Run(
 		ctx,
 		"git",
-		command.RunWithArgs("init", "--bare"),
```
Comment (emcfarlane) [inline private/pkg/git/cloner.go:121]: Removed `--bare` and the below `worktree` in favour of simply checking out the repository into this directory. Are there any requirements for the worktree flow?

File: private/pkg/git/cloner.go [hunk 2]
```diff
@@ -154,97 +132,102 @@ func (c *cloner) CloneToBucket(
...
-		"worktree",
-		"add",
-		worktreeDir.AbsPath(),
-		worktreeRef,
-	)
 	if err := c.runner.Run(
 		ctx,
 		"git",
-		command.RunWithArgs(args...),
+		command.RunWithArgs("checkout", "--force", "FETCH_HEAD"),
```
Comment (emcfarlane) [inline private/pkg/git/cloner.go:185]: `--force` was added as a safeguard following github actions checkout.

File: private/pkg/git/cloner.go [hunk 3]
```diff
@@ -344,41 +327,39 @@ func getSSHKnownHostsFilePaths(sshKnownHostsFiles string) []string {
...
-		// After fetch we won't have checked out any refs. This
-		// will cause `refs=` containing "HEAD" to fail, as HEAD
-		// is a special case that is not fetched into a ref but
-		// instead refers to the current commit checked out. By
-		// checking out "FETCH_HEAD" before checking out the
-		// user supplied ref, we behave similarly to git clone.
-		return "HEAD", "FETCH_HEAD", gitName.checkout()
-	} else {
-		return "HEAD", "FETCH_HEAD", ""
+	checkout, cloneBranch := gitName.checkout(), gitName.cloneBranch()
```
Comment (emcfarlane) [inline private/pkg/git/cloner.go:339]: `cloneBranch` is misnamed, this can be a branch, tag or commit.

File: private/pkg/git/cloner.go (hunk 1 above)
Comment (emcfarlane) [reply private/pkg/git/cloner.go:121]: Need to check this is supported: https://github.com/bufbuild/buf/pull/421

**Review by bufdev — APPROVED** [review]
I am very out of date on this package, so assume I approve, and get a review from someone else more familiar with git.

File: private/pkg/git/cloner.go (hunk 1 above)
Comment (doriable) [reply private/pkg/git/cloner.go:121]: Just tested this behaviour, it seems that this does break #419. Let's chat about this and figure something out.

**Comment by doriable** [comment]
We cannot currently merge this since it breaks the behaviour for #419 -- we'll need to investigate a solution that continues to work with that.

File: private/pkg/git/cloner.go (hunk 1 above)
Comment (emcfarlane) [reply private/pkg/git/cloner.go:121]: I tested on this repo with the following and the commands work as intended:
```bash
$ buf breaking --against 'https://github.com/bufbuild/buf.git#branch=main,ref=HEAD~1'
$ buf breaking --against '.git#branch=main,ref=HEAD~1'
```
For the issue #2463 I created the following `pre-commit` hook:
```
repos:
  - repo: https://github.com/bufbuild/buf
    rev: ed/gitRefResolve
    hooks:
      - id: buf-breaking
        args: ['--against', '.git#branch=main,recurse_submodules=true']
```
which when creating a breaking change seems to be correct:
```
$ git commit
[WARNING] The 'rev' field of repo 'https://github.com/bufbuild/buf' appears to be a mutable reference (moving tag / branch).  Mutable references are never updated after first install and are not supported.  See https://pre-commit.com/#using-the-latest-version-for-a-repository for more details.  Hint: `pre-commit autoupdate` often fixes this.
buf breaking.............................................................Failed
- hook id: buf-breaking
- exit code: 100

proto/acme/example/v1/book.proto:8:1:Previously present field "4" with name "read_status" on message "Book" was deleted.
```
@doriable what was the failure cases you saw?

Comment (doriable) [reply private/pkg/git/cloner.go:121]: So when running the case with `.git#branch=main,ref=HEAD~1`, I get the following (scrubbed the full path of the `/private` directory`):

```
$ buf breaking --against '.git#branch=main,ref=HEAD~1'
Failure: could not clone file:///Users/dkeung/src/buf/buf/.git: exit status 128
fatal: refusing to fetch into branch 'refs/heads/main' checked out at '/private/var/folders/...'
```

Comment (doriable) [reply private/pkg/git/cloner.go:121]: Commenting to keep track of offline discussion: seems like this has to do with the version of Git I had installed, `git version 2.39.3 (Apple Git-146)`. When I swapped to a later version installed through Homebrew, `git version 2.45.2`, it works. We'll need to investigate this for versions of Git.

Comment (emcfarlane) [reply private/pkg/git/cloner.go:121]: Workaround to add `--update-head-ok`. This looks only to be needed by Apple versions of `git`. Testing against `git` v2.39.3 installed from source is fine. But the corresponding `git version 2.39.3 (Apple Git-146)` causes this failure.

Comment (emcfarlane) [reply private/pkg/git/cloner.go:121]: Reason for failure and noticing it only with apple git versions is due to the refspec. Apple git defaults to `main` while normal `git` init branch is master. The refspec of `+main:main` will cause a failure if the current branch is `main` on `git fetch` due to updating the `HEAD` to main, which fetch intentionally disallows. To fix this we now add the `--update-head-ok`, this behaviour is desired and intentional.

**Review by doriable — APPROVED** [review]

**Comment by marcus-griep-simplisafe** [comment]
Related to #3265.

**Comment by emcfarlane** [comment]
@marcus-griep-simplisafe updated in the other issue. Working on a fix.

**Pull request merged**